src/measure_theory/function/ae_eq_of_integral.lean

```diff
@@ -0,0 +1,486 @@
+import measure_theory.function.strongly_measurable
+import measure_theory.integral.set_integral
+
+/-! # From equality of integrals to equality of functions
+
+This file provides various statements of the general form "if two functions have the same integral
+on all sets, then they are equal almost everywhere".
+The different lemmas use various hypotheses on the class of functions, on the target space or on the
+possible finiteness of the measure.
+
+## Main statements
+
+All results listed below apply to two functions `f,g`, together with two main hypotheses,
+* `f` and `g` are integrable on all measurable sets with finite measure,
+* for all measurable sets `s` with finite measure, `∫ x in s, f x ∂μ = ∫ x in s, g x ∂μ`.
+The conclusion is then `f =ᵐ[μ] g`. The main lemmas are:
+* `ae_eq_of_forall_set_integral_eq_of_sigma_finite`: case of a sigma-finite measure.
+* `ae_fin_strongly_measurable.ae_eq_of_forall_set_integral_eq`: for functions which are
+  `ae_fin_strongly_measurable`.
+* `Lp.ae_eq_of_forall_set_integral_eq`: for elements of `Lp`, for `0 < p < ∞`.
+* `integrable.ae_eq_of_forall_set_integral_eq`: for integrable functions.
+
+For each of these results, we also provide a lemma about the equality of one function and 0. For
+example, `Lp.ae_eq_zero_of_forall_set_integral_eq_zero`.
+
+Generally useful lemmas which are not related to integrals:
+* `ae_eq_zero_of_forall_inner`: if for all constants `c`, `λ x, inner c (f x) =ᵐ[μ] 0` then
+  `f =ᵐ[μ] 0`.
+* `ae_eq_zero_of_forall_dual`: if for all constants `c` in the dual space, `λ x, c (f x) =ᵐ[μ] 0`
+  then `f =ᵐ[μ] 0`.
+
+## TODO(s)
+
+* Extend the results from a Hilbert space to a Banach space. Most lemmas about inner products
+  should be equally valid about the product with an element of the dual and
+  `ae_eq_zero_of_forall_dual` should be used instead of `ae_eq_zero_of_forall_inner`.
+
+-/
+
+open measure_theory topological_space normed_space filter
+
+open_locale ennreal measure_theory
+
+namespace measure_theory
+
+
+section ae_eq_of_forall
+
+variables {α E 𝕜 : Type*} {m : measurable_space α} {μ : measure α} [is_R_or_C 𝕜]
+
+lemma ae_eq_zero_of_forall_inner [inner_product_space 𝕜 E] [second_countable_topology E]
+  {f : α → E} (hf : ∀ c : E, (λ x, (inner c (f x) : 𝕜)) =ᵐ[μ] 0) :
+  f =ᵐ[μ] 0 :=
+begin
+  let s := dense_seq E,
+  have hs : dense_range s := dense_range_dense_seq E,
+  have hf' : ∀ᵐ x ∂μ, ∀ n : ℕ, inner (s n) (f x) = (0 : 𝕜), from ae_all_iff.mpr (λ n, hf (s n)),
+  refine hf'.mono (λ x hx, _),
+  rw [pi.zero_apply, ← inner_self_eq_zero],
+  have h_closed : is_closed {c : E | inner c (f x) = (0 : 𝕜)},
+    from is_closed_eq (continuous_id.inner continuous_const) continuous_const,
+  exact @is_closed_property ℕ E _ s (λ c, inner c (f x) = (0 : 𝕜)) hs h_closed (λ n, hx n) _,
+end
+
+local notation `⟪`x`, `y`⟫` := y x
+
+variables (𝕜)
+
+lemma ae_eq_zero_of_forall_dual [normed_group E] [normed_space 𝕜 E]
+  [second_countable_topology (dual 𝕜 E)]
```

Comment:
isn't it enough to assume that `E` itself is second-countable (which is much weaker, think for instance of `L^1`)?

Sketch of proof. Fix a dense sequence `e_n` in `E`, and choose for each `n` an element `c_n` in the dual of norm `1` with `c_n (e_n) = ||e_n||`, by Hahn-Banach. I claim that the intersection of `ker c_n` is reduced to `0`, wich is enough to conclude. Take a nonzero `x`, then choose `n` with `||x-e_n|| < ||x||<2`. Then `c_n (x) >= c_n (e_n) - ||x||/2 = ||e_n|| - ||x||/2 > 1/2||x|| - ||x||/2 = 0`. Therefore, `c_n (x)` is nonzero, i.e., `x` does not belong to `ker c_n`.

Reply:
Thanks. I will fix this. I am however not very familiar with the relevant parts of the library. Can you point me to files where I can find the tools to construct that sequence and find an element with the required norm condition?

Reply:
Thanks a lot for your new proof!